### PR TITLE
Geom: Point/Vector/Angle Swap

### LINKED ISSUE #1015: Geom: Swap old Point/Vector/Angle with new Point/Vector/Angle implementation in new_geom
### Description of the task
Replace usages of old Point/Vector/Angle with new Point/Vector/Angle in new_geom. This is so that any new work done on other new geom classes can be used and not have differing Point/Vector/Angle implementations.

### Acceptance criteria
- Old geom point.h removed
- Old geom angle.h removed
- All usages of old Point/Vector replaced with new Point/Vector
- All usages of old Angle/AngularVelocity replaced with new Angle/AngularVelocity
- All unit tests still pass

### PR DESCRIPTION
**Pull request by scveloso**
### Description
- Replaced usages of old Point/Vector/Angle with new Point/Vector/Angle
- Point ```isClose(Point other, double dist)``` differed from old behavior which caused tests to fail,  should be consistent now.
- Some of the old usages of Point had to be changed to Vector and vice-versa.
- Angle ```mod()``` was used somewhere but was removed in new geom. Added it back in. 
- Some usages of Point made sense that they be a Point but they used Vector operations which is no longer allowed in new geom - such logic used ```toVector()``` and ```Point(Vector v)```

### Testing Done
All unit tests still pass

### Resolved Issues
resolves #1015

### Length Justification
Point/Vector is used extensively within our codebase, doing this in one PR seems better than doing this in small batches and having two Point/Vector implementations used while development continues.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
@EvanMorcom @jonathanlew @jsn0716 please review.

**Comment by scveloso** [comment]
@MathewMacDougall Sorry about the PR size, not much I can do unless we decide to introduce this in small increments (which will have two versions of Point/Vector used at one time). Maybe we can have one reviewer look at each module?

**Comment by MathewMacDougall** [comment]
Yea the size is justifiable in this case. Everyone should review the whole thing since people will catch different things, but I recommend reviewing in small sessions so you're not staring at code for hours.

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (jsn0716) [inline src/software/backend/simulation/physics_ball_test.cpp]: What's the reason for the change to `1e-7`?

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (jsn0716) [inline src/software/backend/simulation/physics_ball_test.cpp]: Should this be tested with `EXPECT_EQ` instead with the `Vector`'s `==` operator? Seems unnecessarily convoluted to have to convert the two `Vector`s to `Point`s just to compare them. If we want to be able to control the value of `EPSILON` for comparing `Vector`'s, maybe `Vector` class should have its own method for it. Also applies to other places like `physics_field_test.cpp`.

File: src/software/geom/util.cpp
Comment (jsn0716) [inline src/software/geom/util.cpp:107]: If we are going to avoid using shortened names, might as well change this to `distanceSquared` as well? But maybe better kept to another PR, don't want this PR getting any bigger 😅

File: src/software/geom/util.cpp
Comment (jsn0716) [inline src/software/geom/util.cpp:924]: How was the `EPS` determined for geometry utilities? We now have an `EPSILON` defined in `geom_constants.h`, but it has a value much smaller than `EPS`. Either way, I think it would be nice to have all constants in `geom_constants.h`. @MathewMacDougall @garethellis0

**Review by jsn0716 — CHANGES_REQUESTED** [review]
Looking good, left some minor comments.

File: src/software/ai/hl/stp/play/corner_kick_play.cpp
Comment (scveloso) [reply src/software/ai/hl/stp/play/corner_kick_play.cpp]: Leftover brackets from when I was using + - and Point Vector subtraction wasn't implemented yet. Fixed in latest update.

File: src/software/ai/hl/stp/play/corner_kick_play.cpp
Comment (scveloso) [reply src/software/ai/hl/stp/play/corner_kick_play.cpp]: Fixed in latest

File: src/software/backend/output/grsim/command_primitive_visitor/motion_controller.cpp
Comment (scveloso) [reply src/software/backend/output/grsim/command_primitive_visitor/motion_controller.cpp]: Ah yes, fixed in latest

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (scveloso) [reply src/software/backend/simulation/physics_ball_test.cpp]: Leftover from different ```isClose()``` definition. Changed back in latest update

Comment (scveloso) [reply src/software/backend/simulation/physics_ball_test.cpp]: Leftover from different ```Point::isClose()``` definition. Changed back to ```1e-7``` in latest update

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (scveloso) [reply src/software/backend/simulation/physics_ball_test.cpp]: Yeah I ran into several of these cases, maybe something like ```Vector::almostEqual(const Vector& other)``` since I'm not sure if it makes sense for two vectors to be "close" to one another.

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp]: Agree, this could be fixed when we port over ```geom/util.cpp``` to ```new_geom```

Comment (scveloso) [reply src/software/geom/util.cpp:107]: Agree, this could be fixed when we adapt ```geom/util.cpp``` to ```new_geom```

File: src/software/geom/util.cpp
Comment (garethellis0) [reply src/software/geom/util.cpp:924]: Here's a good article on what epsilon is and how to determine it: https://en.wikibooks.org/wiki/Floating_Point/Epsilon

Not sure why I never thought to look for it before, but C++ has a `epsilon` in the standard library: https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon, perhaps we should just use this instead?

File: src/software/geom/rectangle.cpp
Comment (garethellis0) [inline src/software/geom/rectangle.cpp]: why not `-=` here?

File: src/software/ai/navigator/obstacle/obstacle_test.cpp
Comment (garethellis0) [inline src/software/ai/navigator/obstacle/obstacle_test.cpp:16]: (out-of-scope, optiona): while we're here, `centred`->`centered`
``` suggestion
    // centered at (0,0) and with (0,0) velocity
```

Comment (garethellis0) [reply src/software/ai/navigator/obstacle/obstacle_test.cpp:16]: Oh my god, this misspelling is _everywhere_, nevermind, we can make this change all at once later...........

File: src/software/geom/util.cpp
Comment (garethellis0) [inline src/software/geom/util.cpp]: Do we only allow one ordering of the `+` operator?

File: src/software/geom/util.cpp
Comment (garethellis0) [inline src/software/geom/util.cpp]: Should this still be `+ -(...`?

File: src/software/new_geom/angle.h
Comment (garethellis0) [inline src/software/new_geom/angle.h:114]: Do we have tests for this?

File: src/software/new_geom/angle.h
Comment (garethellis0) [inline src/software/new_geom/angle.h:454]: I'm confused what the `static_cast<long>` is for here?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good! Not many comments given the scope/size of the PR (it also broke github for me, there's so much lag typing this right now :laughing:)

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball_test.cpp]: Yes I believe the simulation stores things internally as floats and accuracy is lost during the simulation steps (because things can't always be solved perfectly), which is why I had do do custom comparisons like this.

`almostEqual` or a vector equivalent to `isClose` makes sense.

File: src/software/geom/util.cpp
Comment (MathewMacDougall) [reply src/software/geom/util.cpp:924]: Oh neat didn't realize `numeric_limits` had an `epsilon`. Don't know why `geom/util` has its own `EPS` value.

File: src/software/backend/input/network/filter/ball_filter.cpp
Comment (MathewMacDougall) [reply src/software/backend/input/network/filter/ball_filter.cpp]: I think normalizing a vector to change it's length / get unit vectors is valid?

File: src/software/geom/util.cpp
Comment (jsn0716) [reply src/software/geom/util.cpp:924]: So the `epsilon` in standard library is `2.22045e-16`, which is quite a bit smaller than `EPSILON` as defined in `geom_constants.h`. According to its comments: `1e-15 was chosen as a value because doubles have about 16 consistent significant figures. Comparing numbers with 15 significant figures gives us a small buffer while remaining as accurate as possible.` I'm all for using a constant defined by the standard library, but concerned that the EPSILON value provided might be too small. Thoughts?

File: src/software/new_geom/angle.h
Comment (scveloso) [reply src/software/new_geom/angle.h]: Added tests in the latest update

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp]: This was leftover from when I didn't have the new operator. Fixed in latest update

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp]: Yeah I figured only adding a Vector to a Point makes sense rather than adding a Point to a Vector

File: src/software/geom/rectangle.cpp
Comment (scveloso) [reply src/software/geom/rectangle.cpp]: Added a -= operator, fixed in latest update

File: src/software/backend/input/network/filter/ball_filter.cpp
Comment (scveloso) [reply src/software/backend/input/network/filter/ball_filter.cpp]: Also, this is ```norm()``` from ```Eigen```, not our implementation. That comment change is refactoring going too far and I undid it in the latest update

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (scveloso) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Ah yes, fixed ```calcBlockCone(...)``` with 4 parameters to be changed to ```Point```s instead in latest update.

File: src/software/ai/navigator/obstacle/obstacle_test.cpp
Comment (scveloso) [reply src/software/ai/navigator/obstacle/obstacle_test.cpp:16]: Agree, perhaps a different PR for this one.

File: src/software/new_geom/angle.h
Comment (scveloso) [reply src/software/new_geom/angle.h:454]: I'm not sure why either. I just copied this from ```old_geom```. Do you know why @MathewMacDougall?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Wow this is huge!

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (garethellis0) [reply src/software/backend/simulation/physics_ball_test.cpp]: I dislike both `almostEqual` and `isClose`, because they are both pretty subjective. I would be fine with implementing the `==` operator to check if two vectors are equal to within machine precision though? If we want to check if something is close to within `float` tolerance, we should check if both the `x` and `y` components are equal using `std::numeric_limits<float>::epsilon`

File: src/software/geom/util.cpp
Comment (garethellis0) [reply src/software/geom/util.cpp:924]: Hmm, good point. Let's chat about this at meeting tmw.

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: I'm confused, seems like `calcBlockCone` is not in `goalie_tactic.cpp`?

File: src/software/ai/navigator/obstacle/obstacle_test.cpp
Comment (garethellis0) [reply src/software/ai/navigator/obstacle/obstacle_test.cpp:16]: oh interesting, TIL. this seems fine then

File: src/software/geom/util.cpp
Comment (garethellis0) [reply src/software/geom/util.cpp]: Hmm, I'd be inclined to have both, just because addition is usually commutative (order doesn't matter). @MathewMacDougall - second thoughts on this?

File: src/software/new_geom/angle.h
Comment (garethellis0) [reply src/software/new_geom/angle.h:454]: If it's just copy-over then that's fine, out-of-scope for this PR, falls under the geom utils redo!

Comment (MathewMacDougall) [reply src/software/new_geom/angle.h:454]: Not sure either, sorry :man_shrugging:

File: src/software/geom/util.cpp
Comment (MathewMacDougall) [reply src/software/geom/util.cpp]: Hmmm that's a tough call. I think it's more readable if it's _not_ commutative, and you can only add a vector to a point, but technically Gareth is right about addition generally being commutative. I think the main question is only allowing one option more readable and worth the slight extra overhead writing the code?

@jonathanlew @EvanMorcom for their opinions.

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (MathewMacDougall) [reply src/software/backend/simulation/physics_ball_test.cpp]: @garethellis0 for these tests then do you think we should subtract the vectors and take the difference (rather than use a function that determines "closeness")?

Eg. `EXPECT_LT((ball_parameter.velocity() - ball.velocity()).len(), 1e-6);`

File: src/software/backend/input/network/filter/ball_filter.cpp
Comment (MathewMacDougall) [reply src/software/backend/input/network/filter/ball_filter.cpp]: Ah, this should be fine as is then.

File: src/software/new_geom/angle.h
Comment (garethellis0) [reply src/software/new_geom/angle.h:454]: Ok, let's leave this as work for the geom util rewrite. @scveloso - please mark this comment as resolved once you've seen it.

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (garethellis0) [reply src/software/backend/simulation/physics_ball_test.cpp]: Yeah, I'm totally fine with that, main issue I had was having arbitrary methods on the `Point`/`Vector` classes themselves to check "almost equal" or "is close"

File: src/software/geom/util.cpp
Comment (EvanMorcom) [reply src/software/geom/util.cpp]: I think either Vector + Point and Point + Vector make sense intuitively. 

You can have a vector and want to shift it by a Point, or a Point and want a Vector towards something.

File: src/software/new_geom/angle.h
Comment (EvanMorcom) [reply src/software/new_geom/angle.h:454]: Probably sound kind of rounding?

**Review by EvanMorcom — COMMENTED** [review]
LGTM Pending the existing comments :100:

File: src/software/geom/rectangle.h
Comment (scveloso) [reply src/software/geom/rectangle.h]: ~~Division operator is only available for vectors, point + vector results in a point so i dont think we can do that~~ nevermind i misread your comment, will be fixed in a later commit

File: src/software/geom/rectangle.h
Comment (scveloso) [reply src/software/geom/rectangle.h]: fixed in latest update

File: src/software/geom/ray.h
Comment (scveloso) [reply src/software/geom/ray.h]: fixed in latest update

File: src/software/geom/ray.h
Comment (scveloso) [reply src/software/geom/ray.h]: 1. Maybe we should hold off on javadocs as this will be replaced by new_geom ray
2. Fixed in latest update

File: src/software/new_geom/angle.h
Comment (scveloso) [reply src/software/new_geom/angle.h:114]: Got lazy towards the end of the big pr but yeah, added tests in latest update

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp]: Thanks for the review - added commutative addition for Points and Vectors in latest update.

File: src/software/backend/simulation/physics_ball_test.cpp
Comment (scveloso) [reply src/software/backend/simulation/physics_ball_test.cpp]: Changed tests that convert ```Vector```s to ```Point```s THEN uses ```isClose(...)``` to comparing the vector difference to a small length instead (refer to Mathew's example above). This is in the latest update.

File: src/software/geom/rectangle.h
Comment (scveloso) [reply src/software/geom/rectangle.h]: fixed in latest update

**Review by jonathanlew — APPROVED** [review]
LGTM other than what others have brought up

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp:924]: Chatted about it and we now have this issue #1023 feel free to edit if I missed anything

**Review by garethellis0 — APPROVED** [review]
LGTM aside from other peoples comments! :+1: :tada:

**Comment by MathewMacDougall** [comment]
I think we just need @jsn0716 's approval, everyone else's comments look resolved. I have nothing to add beyond what's been said, so should be good to go after that.

File: src/software/geom/util.cpp
Comment (jsn0716) [reply src/software/geom/util.cpp:107]: Added as criteria for #959

**Review by jsn0716 — APPROVED** [review]
LGTM 💯

**Review by MathewMacDougall — APPROVED** [review]
:tada:

**Pull request merged**
